### PR TITLE
harden ClusterShardingPersistenceSpec by using onPostStop, #26230

### PR DESCRIPTION
**Pull request by patriknw**
Also adding onPostStop and onPreRestart hooks to EventSourcedBehavior

Refs #26230
Refs #26404

Debug logging revealed that it was a race condition between sending the Passivate to the shard and processing of next command.

```
[DEBUG] [02/18/2019 13:43:09.777] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Request shard [51] home. Coordinator [Some(Actor[akka://ClusterShardingPersistenceSpec/system/sharding/testCoordinator/singleton/coordinator#50385039])]
[DEBUG] [02/18/2019 13:43:09.778] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/replicator] Received Update for key [testCoordinatorState]
[DEBUG] [02/18/2019 13:43:09.779] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/testCoordinator/singleton/coordinator] The coordinator state was successfully updated with ShardHomeAllocated(51,Actor[akka://ClusterShardingPersistenceSpec/system/sharding/test#1485895496])
[DEBUG] [02/18/2019 13:43:09.779] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/testCoordinator/singleton/coordinator] Shard [51] allocated at [Actor[akka://ClusterShardingPersistenceSpec/system/sharding/test#1485895496]]
[DEBUG] [02/18/2019 13:43:09.779] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Host Shard [51] 
[DEBUG] [02/18/2019 13:43:09.780] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Starting shard [51] in region
[DEBUG] [02/18/2019 13:43:09.781] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Shard [51] located at [Actor[akka://ClusterShardingPersistenceSpec/system/sharding/test#1485895496]]
[DEBUG] [02/18/2019 13:43:09.781] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Shard was initialized 51
[DEBUG] [02/18/2019 13:43:09.781] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-22] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test] Deliver [1] buffered messages for shard [51]
[DEBUG] [02/18/2019 13:43:09.782] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test/51] Starting entity [3] in shard [51]
[DEBUG] [02/18/2019 13:43:09.786] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-2] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:get-permit] Initializing snapshot recovery: Recovery(SnapshotSelectionCriteria(9223372036854775807,9223372036854775807,0,0),9223372036854775807,9223372036854775807)
[DEBUG] [02/18/2019 13:43:09.788] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:replay-evts] Replaying messages: from: 1, to: 9223372036854775807
[DEBUG] [02/18/2019 13:43:09.789] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:replay-evts] Recovery successful, recovered until sequenceNr: [0]
[DEBUG] [02/18/2019 13:43:09.790] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:replay-evts] Returning recovery permit, reason: replay completed successfully
[DEBUG] [02/18/2019 13:43:09.790] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3] onRecoveryCompleted: []

>>> 1
[DEBUG] [02/18/2019 13:43:09.792] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:running-cmnds] Handled command [akka.cluster.sharding.typed.scaladsl.ClusterShardingPersistenceSpec$PassivateAndPersist], resulting effect: [Persist(1)], side effects: [1]
[DEBUG] [02/18/2019 13:43:09.793] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-14] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:persist-evts] Received Journal response: WriteMessagesSuccessful
[DEBUG] [02/18/2019 13:43:09.794] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-14] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:persist-evts] Received Journal response: WriteMessageSuccess(PersistentImpl(1,1,test|3,,false,null,cffad868-592e-4a2f-ae10-d7676b5bec62),3)
[DEBUG] [02/18/2019 13:43:09.795] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-2] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:running-cmnds] 

>>> 2
Handled command [akka.cluster.sharding.typed.scaladsl.ClusterShardingPersistenceSpec$PassivateAndPersist], resulting effect: [Persist(2)], side effects: [1]

>>>
[DEBUG] [02/18/2019 13:43:09.795] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-14] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test/51] Passivation already in progress for Actor[akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3#367017175]. Not sending stopMessage back to entity.
[DEBUG] [02/18/2019 13:43:09.798] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-2] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:persist-evts] Received Journal response: WriteMessagesSuccessful
[DEBUG] [02/18/2019 13:43:09.798] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-2] [akka://ClusterShardingPersistenceSpec/system/sharding/test/51/3][persistenceId:test|3][phase:persist-evts] Received Journal response: WriteMessageSuccess(PersistentImpl(2,2,test|3,,false,null,cffad868-592e-4a2f-ae10-d7676b5bec62),3)
[DEBUG] [02/18/2019 13:43:09.806] [ClusterShardingPersistenceSpec-akka.actor.default-dispatcher-23] [akka://ClusterShardingPersistenceSpec@127.0.0.1:60240/system/sharding/test/51] Entity stopped after passivation [3]
[info] - must handle PoisonPill after persist effect *** FAILED *** (10 seconds, 33 milliseconds)
[info]   java.lang.AssertionError: timeout (10 seconds) during expectMessage while waiting for recoveryCompleted:1
```

### CODE REVIEW

File: akka-actor/src/main/scala/akka/util/ConstantFun.scala [hunk 1]
```diff
@@ -44,6 +44,8 @@ import akka.japi.{ Pair ⇒ JPair }
 
   val oneInt = (_: Any) ⇒ 1
 
+  val unitToUnit = () ⇒ ()
```
Comment (chbatey) [inline akka-actor/src/main/scala/akka/util/ConstantFun.scala]: best type of fun

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala [hunk 2]
```diff
@@ -36,7 +36,7 @@ import org.scalatest.WordSpecLike
 object ClusterShardingPersistenceSpec {
   val config = ConfigFactory.parseString(
     """
-      akka.loglevel = INFO
+      akka.loglevel = DEBUG
```
Comment (chbatey) [inline akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: I guess switch this back or use the fancy logger that only outputs on failure?

**Review by chbatey — APPROVED** [review]
LGTM

File: akka-actor/src/main/scala/akka/util/ConstantFun.scala (hunk 1 above)
Comment (helena) [reply akka-actor/src/main/scala/akka/util/ConstantFun.scala]: I think I needed this recently

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala [hunk 3]
```diff
@@ -128,11 +128,17 @@ object ClusterShardingPersistenceSpec {
...
-            case null ⇒ ctx.log.debug("no recoveryCompletedProbe for [{}]", entityId)
+          lifecycleProbes.get(entityId) match {
+            case null ⇒ ctx.log.debug("no lifecycleProbe (onRecoveryCompleted) for [{}]", entityId)
             case p    ⇒ p ! s"recoveryCompleted:$state"
           }
         }
+        .onPostStop(() ⇒
+          lifecycleProbes.get(entityId) match {
+            case null ⇒ ctx.log.debug("no lifecycleProbe (postStop) for [{}]", entityId)
+            case p    ⇒ p ! s"stopped"
```
Comment (helena) [inline akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]:
```suggestion
            case p    ⇒ p ! "stopped"
```

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala [hunk 4]
```diff
@@ -460,22 +478,22 @@ class ClusterShardingPersistenceSpec extends ScalaTestWithActorTestKit(ClusterSh
...
       ref ! Add("3")
       ref ! Add("4")
       ref ! UnstashAllAndPassivate
 
       val probe = TestProbe[String]()
-      recoveryProbe.awaitAssert {
+      lifecycleProbe.awaitAssert {
         ref ! Get(probe.ref)
-        recoveryProbe.expectMessage(max = 1.second, "recoveryCompleted:" + (1 to 4).map(_.toString).mkString("|"))
+        lifecycleProbe.expectMessage(max = 1.second, "recoveryCompleted:" + (1 to 4).map(_.toString).mkString("|"))
```
Comment (helena) [inline akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/scaladsl/ClusterShardingPersistenceSpec.scala]: It looks like we do this op 2x so why not
```scala
val expected = (1 to 4).map(_.toString).mkString("|")
lifecycleProbe.expectMessage(max = 1.second, "recoveryCompleted:" + $expected)
probe.expectMessage(entityId + ":" + $expected)
```

**Review by helena — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
incorporated feedback, will merge after validation

**Pull request merged**